radix-transaction-scenarios/src/executor.rs

```diff
@@ -0,0 +1,592 @@
+use radix_transactions::errors::*;
+use radix_transactions::validation::*;
+use sbor::prelude::*;
+
+use scenarios::account_authorized_depositors::AccountAuthorizedDepositorsScenarioCreator;
+use scenarios::account_locker::AccountLockerScenarioCreator;
+use scenarios::fungible_resource::FungibleResourceScenarioCreator;
+use scenarios::global_n_owned::GlobalNOwnedScenarioCreator;
+use scenarios::kv_store_with_remote_type::KVStoreScenarioCreator;
+use scenarios::max_transaction::MaxTransactionScenarioCreator;
+use scenarios::metadata::MetadataScenario;
+use scenarios::non_fungible_resource::NonFungibleResourceScenarioCreator;
+use scenarios::non_fungible_resource_with_remote_type::NonFungibleResourceWithRemoteTypeScenarioCreator;
+use scenarios::radiswap::RadiswapScenarioCreator;
+use scenarios::royalties::RoyaltiesScenarioCreator;
+use scenarios::transfer_xrd::TransferXrdScenarioCreator;
+
+type ScenarioCreatorDynFunc = dyn FnOnce(ScenarioCore) -> Box<dyn ScenarioInstance>;
+
+pub struct TransactionScenarioExecutor<D, W, E, F1, F2, F3, F4>
+where
+    D: SubstateDatabase + CommittableSubstateDatabase,
+    W: WasmEngine,
+    E: NativeVmExtension,
+    F1: FnMut(&NetworkDefinition, ProtocolUpdate, &mut D),
+    F2: FnMut(&ScenarioMetadata, &NextTransaction, &TransactionReceiptV1, &D),
+    F3: FnMut(&ScenarioMetadata),
+    F4: FnMut(&NetworkDefinition, &mut D),
+{
+    /* Environment */
+    /// The substate database that the scenario will be run against.
+    database: D,
+    /// The Scrypto VM to use in executing the scenarios.
+    scrypto_vm: ScryptoVm<W>,
+    /// The Native VM to use in executing the scenarios.
+    native_vm_extension: E,
+
+    /* Execution */
+    /// A filter that controls which scenarios are executed and which are not. When [`None`] then
+    /// all scenarios are executed and when [`Some`] then some scenarios may be filtered out.
+    filter: Option<ScenarioFilter>,
+    /// A map of the scenarios registered on the executor. Not all registered scenarios will be
+    /// executed, it merely informs the executor of the existence of these scenarios. Execution of a
+    /// scenario requires that is passes the filter specified by the client.
+    registered_scenarios: BTreeMap<Option<ProtocolUpdate>, Vec<Box<ScenarioCreatorDynFunc>>>,
+    /// Controls whether the bootstrap process should be performed or not.
+    bootstrap: bool,
+    /// The first nonce to use in the execution of the scenarios.
+    starting_nonce: u32,
+    /// How the executor should handle nonces and how it should get the next nonce.
+    next_scenario_nonce_handling: ScenarioStartNonceHandling,
+    /// The network definition to use in the execution of the scenarios.
+    network_definition: NetworkDefinition,
+
+    /* Callbacks */
+    /// A callback that is called when a new protocol requirement is encountered. This can be useful
+    /// for clients who wish to apply protocol updates they wish.
+    on_new_protocol_requirement_encountered: F1,
```

Comment:
As mentioned, if we make these `Box<dyn FnMut(...)>` I imagine the code would be neater and easier to work with :)

Reply:
I typically prefer not to use dynamic dispatch when generics can be used but this seems like a perfect use case for dynamic dispatch. As more and more callbacks were added it was getting somewhat difficult to do the struct updates, which dynamic dispatch solved beautifully. Additionally, I do not suspect that the performance loss from this would be noticable since this part of scenario execution is not performance criticial.

It feels like something could also be done here with macros also or perhaps even a more declarative way to do [Builders with Type State](https://www.youtube.com/watch?v=k8kd22jNcps), but perhaps something that we can exlore later.